src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,958 @@
+                                  std::is_same<BsonT, bsoncxx::types::b_int64>::value ||
+                                  std::is_same<BsonT, bsoncxx::oid>::value;
+};
+
+/**
+ * A templated struct containing a bool value that specifies whether the
+ * provided template parameter is a BSON type that contains a view.
+ */
+template <class BsonT>
+struct is_bson_view {
+    static constexpr auto value = std::is_same<BsonT, bsoncxx::types::b_utf8>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_document>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_array>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_binary>::value;
+};
+
+class BSONOutputArchive : public OutputArchive<BSONOutputArchive> {
+    /**
+    * The possible states for the BSON nodes being output by the archive.
+    */
+    enum class OutputNodeType { StartObject, InObject, StartArray, InArray };
+
+    typedef bsoncxx::builder::core BSONBuilder;
+
+   public:
+    /**
+    * Construct a BSONOutputArchive that will output to the provided stream.
+    *
+    * @param stream
+    *   The stream to which the archiver will output BSON data.
+    */
+    BSONOutputArchive(std::ostream& stream)
+        : OutputArchive<BSONOutputArchive>{this},
+          _bsonBuilder{false},
+          _writeStream{stream},
+          _nextName{nullptr} {
+    }
+
+   private:
+    /**
+     * Writes the current contents of the BSON document builder to the
+     * output stream.
+     */
+    void writeDoc() {
+        _writeStream.write(reinterpret_cast<const char*>(_bsonBuilder.view_document().data()),
+                           _bsonBuilder.view_document().length());
+        _bsonBuilder.clear();
+    }
+
+   public:
+    /**
+     * Checks if the most recent object written was in the root of the document.
+     * If so, that document is written to the archive.
+     */
+    void writeDocIfRoot() {
+        if (_nodeTypeStack.empty()) {
+            writeDoc();
+        }
+    }
+
+    /**
+     * Starts a new node in the BSON output.
+     *
+     * The node can optionally be given a name by calling setNextName prior to
+     * creating the node. Nodes only need to be started for types that are
+     * objects or arrays.
+     */
+    void startNode() {
+        writeName(true);
+        _nodeTypeStack.push(OutputNodeType::StartObject);
+    }
+
+    /**
+     * Designates the most recently added node as finished, which will end
+     * any objects or arrays we happen to be in.
+     */
+    void finishNode() {
+        // Iff we ended up serializing an empty object or array, writeName
+        // will never have been called - therefore the node type at the top
+        // of the stack will be StartArray or StartObject - so start and
+        // then immediately end the documents/array.
+
+        // We'll also end any documents/arrays we happen to be in.
+
+        bool closedObj = false;
```

Comment:
The boolean is no longer necessary, it was an artifact from when I had Root as a NodeType. Theoretically _nodeTypeStack.empty() should never be true but I'll add a check just so we don't segfault.